### DISCUSSION TITLE
Auto infer generic augment from target type

### DISCUSSION THREAD

**Discussion by hez2010** [dotnet/csharplang#6272] [Language Ideas]
Consider below snippet:

```csharp
V Add<T, U, V>(T l, U r)
    where T : IAdditionOperators<T, U, V>
    where U : IAdditionOperators<T, U, V>
{
    return l + r;
}
```

When I call this method, I have to specify all generic parameters like:

```csharp
int v = Add<int, int, int>(1, 2, 3);
```

Instead, when target type is known, the compiler should infer `V` from the usage so that we can just write:

```csharp
int v = Add(1, 2, 3);
```

**Comment by ufcpp** [comment]
Related to https://github.com/dotnet/csharplang/issues/339 https://github.com/dotnet/csharplang/issues/1349

**Comment by iam3yal** [comment]
@hez2010 The compiler shouldn't try and infer the generic type arguments based on the type of the variable at the left side, it should do that based on the actual arguments you pass to the method.

**Comment by svick** [comment]
This looks like a duplicate of https://github.com/dotnet/csharplang/discussions/997 and https://github.com/dotnet/csharplang/discussions/741.
